07.ls_object/lib/list.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+require_relative 'file_data'
+
+class List
+  attr_reader :file_datas
+
+  def initialize(paths)
+    @file_datas = paths.map { |path| FileData.new(path) }
+  end
+
+  def max_file_name_width
+    @file_datas.map { |file_data| file_data.filename.size }.max
+  end
+
+  def max_size_width
+    @file_datas.map { |file_data| file_data.size.to_i }.max.to_s.size
+  end
+
+  def max_nlink_width
+    @file_datas.map { |file_data| file_data.nlink.to_i }.max.to_s.size
+  end
+
+  def max_user_width
+    @file_datas.map { |file_data| file_data.user_name.size }.max
+  end
+
+  def max_group_width
+    @file_datas.map { |file_data| file_data.group_name.size }.max
+  end
```

Comment:
たとえば、ですが、こんなふうにsleepを挟んでプログラムを実行してみてください。
めちゃくちゃ遅くなりませんか？

```suggestion
  def max_file_name_width
    @file_datas.map { |file_data| sleep 0.5; file_data.filename.size }.max
  end

  def max_size_width
    @file_datas.map { |file_data| sleep 0.5; file_data.size.to_i }.max.to_s.size
  end

  def max_nlink_width
    @file_datas.map { |file_data| sleep 0.5; file_data.nlink.to_i }.max.to_s.size
  end

  def max_user_width
    @file_datas.map { |file_data| sleep 0.5; file_data.user_name.size }.max
  end

  def max_group_width
    @file_datas.map { |file_data| sleep 0.5; file_data.group_name.size }.max
  end
```

最近のマシンは速いので多少効率が悪いプログラムを書いてもぱっと動きますが、業務レベルになるとチリツモでパフォーマンスを悪化させることがあります。
なぜ効率が悪いのかを考えて、改善してみてほしいです〜。

Reply:
ありがとうございます。
最初は見当違いのところを考えて迷走しておりました。（桁数計算に`Math.log10`使うのかな？等💦）

最大値を求める計算は１度だけでいいのに、ループを回すたびに行っていました。
また、`build_data()`と`format_row()`で2回文字幅調整を行っていました。
1回だけになるように修正しました。